core/src/main/java/org/apache/iceberg/deletes/BitmapPositionDeleteIndex.java

```diff
@@ -40,7 +46,11 @@ public void delete(long posStart, long posEnd) {
 
   @Override
   public boolean isDeleted(long position) {
-    return roaring64Bitmap.contains(position);
+    boolean posIsDeleted = roaring64Bitmap.contains(position);
+    if (counter != null && posIsDeleted) {
+      counter.increment();
+    }
```

Comment:
Given that we have defined the `DeleteCounter` interface ourselves, does it make sense to define a default `NullDeleteCounter` that is a no-op?

There are a number of places where we check `if (counter != null)`.

Reply:
Also, is there any concern that this makes calls to `isDeleted` not idempotent?

Reply:
There are existing tests that have no need for counting deletes, that call some static methods in `Deletes`. I kept those forms of the methods and had them call new forms that take a `DeleteCounter`, passing a null `DeleteCounter`. That is the reason why a `DeleteCounter` field might be null. The way the code is now, if we decide to do without `DeleteCounter` and simply use an `AtomicLong`, then it can be updated very easily. I do understand your suggestion for a `NullDeleteCounter`.

You raise a good point about `PositionDeleteIndex#isDeleted` not being idempotent now. I have checked where it is called, and it is not a concern. We should document this point though, to prevent it being a problem in future code. Before I ended up introducing a delete counter into the `PositionDeleteIndex` implementation and incrementing it when `isDeleted` is called, I considered other approaches, but there was one read path in particular where this seemed to the best solution.

Reply:
Added a note to the javadoc for `PositionDeleteIndex#isDeleted`.

Reply:
Shouldn't we be able to determine the number of deleted rows here by checking the number of entries in the bitmap? So shouldn't we be able to just count in the "delete methods"? Also do we have to pass through the counter here? Seems like we could just get this value out of the index when we are done building it instead of passing through the counter into the index?

Reply:
You are right that one can get the number of deleted positions in the `PositionDeleteIndex` by `roaring64Bitmap.getLongCardinality()`.
However, having one counter keeps the bookkeeping simple and easy to reason about.
There are multiple paths through the code that can be taken, where deletes may be applied, and the unifying thing here is that the same counter is passed into all these places.
It is more difficult to reason through the counting if I have to account for the `PositionDeleteIndex` cases separately and I would have to keep references to the `PositionDeleteIndex` instances (I haven't looked in depth into this). I think it would be messy.

Reply:
I think having "isDeleted" have a side effect ruins the book keeping simplicity for me. I'd much rather if we were to do this counter approach we would the "delete" methods rather than the "isDeleted" method. The delete method already has a side effect and we would expect it to be the place where the count would happen as opposed to "isDeleted" which we have to add a JavaDoc note to for illustrating the side effect behavior. 

I still am not really sold on passing through the counter here, it just seems to me like we are complicating the internals of this class and we don't really need to since like we've already discussed, the value we are looking for is already calculated even without us passing through  an object. I'm also not a big fan of behaviors being dependent on objects being null or not.

But if other reviewers feel differently I could go along with this approach as well.

Reply:
Since it has been almost 4 weeks since I touched this (it seems that's how long the last update has gone unreviewed), I had forgotten the details of what I had investigated and had done. I went through the code paths again, and I see now that there is just one place where `PositionDeleteIndex#delete()` is called and where I need to track the delete count in the `PositionDeleteIndex`. So your suggestion is very feasible and not messy as I feared!
Thank you for the suggestion. I'll update the PR tomorrow.

Reply:
@kbendick I considered using an `Optional<DeleteCounter>` instead of a `DeleteCounter` that could be null, but it's six of one and half a dozen of the other. In the end, I adopted your suggestion to have a no-op `DeleteCounter`.